clang/lib/CIR/Dialect/Transforms/LoweringPrepareArm64CXXABI.cpp

```diff
@@ -0,0 +1,113 @@
+//====- LoweringPrepareArm64CXXABI.cpp - Arm64 ABI specific code --------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provides ARM64 C++ ABI specific code that is used during LLVMIR
+// lowering prepare.
+//
+//===----------------------------------------------------------------------===//
+
+#include "LoweringPrepareItaniumCXXABI.h"
+
+using cir::LoweringPrepareCXXABI;
+using namespace mlir;
+using namespace mlir::cir;
+
+namespace {
+class LoweringPrepareArm64CXXABI : public LoweringPrepareItaniumCXXABI {
+public:
+  mlir::Value lowerVAArg(CIRBaseBuilderTy &builder,
+                         mlir::cir::VAArgOp op) override;
+};
+} // namespace
+
+LoweringPrepareCXXABI *LoweringPrepareCXXABI::createArm64ABI() {
+  return new LoweringPrepareArm64CXXABI();
+}
+
+mlir::Value LoweringPrepareArm64CXXABI::lowerVAArg(CIRBaseBuilderTy &builder,
```

Comment:
We could probably use more comments, but maybe you can reuse some (explained later in reply).
`lowerVAArg` impl approach looks good overall, we just need to frame it more like traditional codegen, as to make sure we track what's missing. This is how you should break this up:

0. Perhaps we can just rename this file `AArch64.cpp`.
1. `lowerVAArg` should start with a body like this below (clang/lib/CodeGen/Targets/AArch64.cpp):

```
Address EmitVAArg(CodeGenFunction &CGF, Address VAListAddr,
                  QualType Ty) const override {
  llvm::Type *BaseTy = CGF.ConvertType(Ty);
  if (isa<llvm::ScalableVectorType>(BaseTy))
    llvm::report_fatal_error("Passing SVE types to variadic functions is "
                             "currently not supported");

  return Kind == AArch64ABIKind::Win64 ? EmitMSVAArg(CGF, VAListAddr, Ty)
         : isDarwinPCS()               ? EmitDarwinVAArg(VAListAddr, Ty, CGF)
                         : EmitAAPCSVAArg(VAListAddr, Ty, CGF, Kind);
}
```

2. Everything not supported becomes an assertion. `ScalableVectorType` stuff could use assert on unimplemented, see examples like `assert(!UnimplementedFeature::somefeaure())`.
3. Then, for `lowerAAPCSVAArg`, you implement what's currently here. Please try to follow the skeleton and reuse lots of the existing comments to clarify approach, explaning, where relevant, the bits that differs and a brief comment on why so.